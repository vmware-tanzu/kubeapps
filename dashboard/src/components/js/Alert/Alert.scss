// Copyright 2020-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

.alert {
  &:not(.alert-app-level) {
    .alert-action {
      height: auto;
      padding: 0;
      border: 0;
      margin: 0;
      background: transparent;
      font-size: 13px;
      font-weight: 400;
      letter-spacing: normal;
      line-height: 1.15;
      text-transform: none;

      &:hover {
        background-color: transparent;
      }
    }
  }
}
