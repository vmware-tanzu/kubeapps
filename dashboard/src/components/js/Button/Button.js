// Copyright 2020-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import cs from "classnames";
import PropTypes from "prop-types";
import React from "react";
import { Link } from "react-router-dom";
import Spinner from "../Spinner";
// Custom Styles
import "./Button.scss";

// Constants
export const ButtonThemes = {
  danger: "danger",
  inverse: "inverse",
  primary: "primary",
  success: "success",
  warning: "warning",
};

export const ButtonTypes = {
  button: "button",
  submit: "submit",
  reset: "reset",
};

/**
 * @see https://clarity.design/documentation/buttons#examples
 */
const Button = ({
  block,
  className,
  children,
  disabled,
  icon,
  externalLink,
  flat,
  link,
  loading,
  loadingText,
  onClick,
  outline,
  rel,
  target,
  theme,
  title,
  type,
  small,
}) => {
  const css = cs("btn button", className, {
    [`btn-${theme}`]: !flat && !outline,
    [`btn-${theme}-outline`]: outline,
    [`btn-link`]: flat,
    [`btn-icon`]: icon,
    [`btn-sm`]: small,
    [`btn-block`]: block,
  });

  if (externalLink != null) {
    // External URL
    return (
      <a
        className={css}
        title={title}
        href={externalLink}
        disabled={disabled}
        rel={rel}
        target={target}
      >
        {children}
      </a>
    );
  } else if (link != null) {
    // Internal URL
    return (
      <Link className={css} title={title} to={link} disabled={disabled}>
        {children}
      </Link>
    );
  } else {
    // Normal Button
    return (
      <button className={css} title={title} onClick={onClick} disabled={disabled} type={type}>
        {loading ? <Spinner inline text={loadingText} /> : children}
      </button>
    );
  }
};

Button.propTypes = {
  block: PropTypes.bool,
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  disabled: PropTypes.bool,
  icon: PropTypes.bool,
  externalLink: PropTypes.string,
  flat: PropTypes.bool,
  link: PropTypes.string,
  loading: PropTypes.bool,
  loadingText: PropTypes.string,
  onClick: PropTypes.func,
  outline: PropTypes.bool,
  rel: PropTypes.string,
  small: PropTypes.bool,
  target: PropTypes.string,
  theme: PropTypes.oneOf(Object.values(ButtonThemes)).isRequired,
  title: PropTypes.string,
  type: PropTypes.oneOf(Object.values(ButtonTypes)).isRequired,
};

Button.defaultProps = {
  block: false,
  className: "",
  disabled: false,
  flat: false,
  icon: false,
  outline: false,
  loading: false,
  small: false,
  theme: ButtonThemes.primary,
  title: "",
  type: ButtonTypes.button,
};

export default Button;
