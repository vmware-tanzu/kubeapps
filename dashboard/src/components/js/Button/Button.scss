// Copyright 2020-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

.button {
  // Override default margin for Clarity
  margin-right: 0 !important;

  & + .button {
    margin-left: 0.5rem;
  }
}
