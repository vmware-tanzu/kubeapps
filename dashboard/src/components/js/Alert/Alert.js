// Copyright 2020-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { CdsIcon } from "@cds/react/icon";
import cs from "classnames";
import PropTypes from "prop-types";
import React from "react";
import Button from "../Button";
import "./Alert.scss";
import AlerTypes from "./Alert.types";

// Constants
export const AlertThemes = {
  danger: "danger",
  info: "info",
  success: "success",
  warning: "warning",
};

export const AlertIcons = {
  danger: "exclamation-circle",
  info: "info-circle",
  success: "check-circle",
  warning: "exclamation-triangle",
};

/**
 * https://v2.clarity.design/alerts
 */
const Alert = ({ app, action, children, customIcon, compact, theme, onClick, onClose }) => {
  // We only want to announce the section on app level alerts
  const accessibilityProps = app
    ? {
        "aria-label": "Please, read the following important information",
      }
    : {};
  const css = cs(`alert alert-${theme}`, {
    "alert-app-level": app,
    "alert-sm": compact,
  });
  // Custom icons are only available for app level
  let icon = app && customIcon != null ? customIcon : AlertIcons[theme];

  // If we provide an action, we should use alertdialog as the role
  // https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Alert_Role
  if (onClose != null || onClick != null) {
    accessibilityProps.role = "alertdialog";
  } else {
    accessibilityProps.role = "alert";
  }

  return (
    <section className={css} {...accessibilityProps}>
      <div className="alert-items">
        <div className="alert-item static">
          <div className="alert-icon-wrapper">
            <CdsIcon class="alert-icon" shape={icon} aria-hidden="true"></CdsIcon>
          </div>
          <div className="alert-text">{children}</div>
          {onClick != null && action != null && (
            <div className="alert-actions">
              <Button className="alert-action" onClick={onClick}>
                {action}
              </Button>
            </div>
          )}
        </div>
      </div>
      {onClose != null && (
        <button type="button" onClick={onClose} className="close" aria-label="Close alert">
          <CdsIcon aria-hidden="true" shape="close"></CdsIcon>
        </button>
      )}
    </section>
  );
};

Alert.propTypes = {
  action: PropTypes.string,
  app: PropTypes.bool,
  children: React.ReactNode || PropTypes.node.isRequired,
  compact: PropTypes.bool,
  customIcon: AlerTypes.customIcon,
  theme: AlerTypes.theme,
  onClick: PropTypes.func,
  onClose: PropTypes.func,
};

Alert.defaultProps = {
  app: false,
  compact: false,
  theme: AlertThemes.info,
};

export default Alert;
