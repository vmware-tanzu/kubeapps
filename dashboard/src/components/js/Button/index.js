// Copyright 2020-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import Button, { ButtonThemes, ButtonTypes } from "./Button";
export { ButtonThemes, ButtonTypes };
export default Button;
