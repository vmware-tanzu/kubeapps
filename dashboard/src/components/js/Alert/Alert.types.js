// Copyright 2020-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { AlertThemes } from "./Alert";

/**
 * Types for alerts
 */

const customIcon = (props, propName, componentName) => {
  const value = props[propName];
  const { app } = props;

  if (app !== true && typeof value === "string" && value.length > 0) {
    return new Error(
      `Invalid ${propName} for ${componentName}. Custom Icon is only available for app level alerts`,
    );
  } else if (value != null && typeof value !== "string") {
    return new Error(
      `Invalid ${propName} for ${componentName}. Custom Icon property must be an String.`,
    );
  } else {
    return null;
  }
};

const theme = (props, propName, componentName) => {
  const value = props[propName];
  const { app } = props;
  const expectedValues = app
    ? Object.values(AlertThemes).filter(v => v !== "success")
    : Object.values(AlertThemes);

  if (expectedValues.includes(value)) {
    return null;
  } else {
    let error = `Invalid ${propName} for ${componentName}. Available values are: ${expectedValues.join(
      ", ",
    )}`;
    if (app && value === "success") {
      error += ". Remember success is not a valid value for App level Alerts.";
    }
    return new Error(error);
  }
};

export default {
  theme,
  customIcon,
};
