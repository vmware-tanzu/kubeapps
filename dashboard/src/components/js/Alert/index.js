// Copyright 2020-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import Alert, { AlertThemes } from "./Alert";
export { AlertThemes };
export default Alert;
