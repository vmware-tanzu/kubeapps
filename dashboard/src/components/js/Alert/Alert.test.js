// Copyright 2020-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { CdsIcon } from "@cds/react/icon";
import { shallow } from "enzyme";
import React from "react";
import Button from "../Button";
import Alert, { AlertIcons, AlertThemes } from "./Alert";

describe(Alert, () => {
  it("renders the required HTML structure", () => {
    const wrapper = shallow(<Alert>Test</Alert>);

    expect(wrapper).toMatchSnapshot();
  });

  it("renders the content without action and close button", () => {
    const text = "test";
    const wrapper = shallow(<Alert>{text}</Alert>);

    expect(wrapper.text()).toContain(text);
    expect(wrapper.find("button")).not.toExist();
    expect(wrapper.find(Button)).not.toExist();
  });

  it("renders a functional action button when the props are present", () => {
    const mock = jest.fn();
    const wrapper = shallow(
      <Alert action="test" onClick={mock}>
        Test
      </Alert>,
    );

    expect(wrapper.find(".alert-action")).toExist();
    wrapper.find(".alert-action").simulate("click");
    expect(mock).toHaveBeenCalled();
  });

  it("renders a functional close button when the props are present", () => {
    const mock = jest.fn();
    const wrapper = shallow(<Alert onClose={mock}>Test</Alert>);

    expect(wrapper.find(".close")).toExist();
    wrapper.find(".close").simulate("click");
    expect(mock).toHaveBeenCalled();
  });

  describe("Accessibility", () => {
    it("adds the alert role when there's no button", () => {
      const wrapper = shallow(<Alert>Test</Alert>);

      expect(wrapper.prop("role")).toBe("alert");
    });

    it("adds the alertdialog role when the notification can be closed", () => {
      const wrapper = shallow(<Alert onClose={() => {}}>Test</Alert>);

      expect(wrapper.prop("role")).toBe("alertdialog");
    });

    it("adds the alertdialog role when there's an action (App alerts)", () => {
      const wrapper = shallow(
        <Alert app action="Action" onClick={() => {}}>
          Test
        </Alert>,
      );

      expect(wrapper.prop("role")).toBe("alertdialog");
    });

    it("includes a label to notify this information is important (App alerts)", () => {
      const wrapper = shallow(<Alert app>Test</Alert>);

      expect(wrapper.prop("aria-label")).toBe("Please, read the following important information");
    });
  });

  describe("Themes", () => {
    Object.values(AlertThemes).forEach(k => {
      const theme = AlertThemes[k];
      const icon = AlertIcons[k];

      it(`apply the ${theme} theme`, () => {
        const wrapper = shallow(<Alert theme={theme}>Test</Alert>);

        expect(wrapper).toHaveClassName(`alert-${theme}`);
        expect(wrapper.find(CdsIcon).prop("shape")).toBe(icon);
      });
    });
  });

  describe("App-level Alerts", () => {
    it("adds the app-level CSS class", () => {
      const wrapper = shallow(<Alert app>Test</Alert>);

      expect(wrapper).toHaveClassName("alert-app-level");
    });

    it("adds a custom icon", () => {
      const customIcon = "objects";
      const wrapper = shallow(
        <Alert app customIcon={customIcon}>
          Test
        </Alert>,
      );

      expect(wrapper.find(CdsIcon).prop("shape")).toBe(customIcon);
    });
  });
});
