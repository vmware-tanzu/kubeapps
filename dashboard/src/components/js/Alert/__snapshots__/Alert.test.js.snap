// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert renders the required HTML structure 1`] = `
<section
  className="alert alert-info"
  role="alert"
>
  <div
    className="alert-items"
  >
    <div
      className="alert-item static"
    >
      <div
        className="alert-icon-wrapper"
      >
        <CdsIcon
          aria-hidden="true"
          class="alert-icon"
          shape="info-circle"
        />
      </div>
      <div
        className="alert-text"
      >
        Test
      </div>
    </div>
  </div>
</section>
`;
